File: VB/DXSample/CustomControls.vb
﻿Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data
Imports System.Windows.Markup
Imports DevExpress.Xpf.Core
Imports DevExpress.Xpf.Core.Internal
Imports DevExpress.Xpf.Utils.Themes

Namespace DXSample
	Public Class CustomControl1
		Inherits Control

		Shared Sub New()
			DefaultStyleKeyProperty.OverrideMetadata(GetType(CustomControl1), New FrameworkPropertyMetadata(GetType(CustomControl1)))
		End Sub

		Public Property Text() As String
			Get
				Return DirectCast(GetValue(TextProperty), String)
			End Get
			Set(ByVal value As String)
				SetValue(TextProperty, value)
			End Set
		End Property
		Public Shared ReadOnly TextProperty As DependencyProperty = DependencyProperty.Register("Text", GetType(String), GetType(CustomControl1), New PropertyMetadata(Nothing))
	End Class

	Public Class CustomControl2
		Inherits CustomControl1

		Shared Sub New()
			DefaultStyleKeyProperty.OverrideMetadata(GetType(CustomControl2), New FrameworkPropertyMetadata(GetType(CustomControl2)))
		End Sub
	End Class

	Public Class CustomControl3
		Inherits CustomControl1

		Shared Sub New()
			DefaultStyleKeyProperty.OverrideMetadata(GetType(CustomControl3), New FrameworkPropertyMetadata(GetType(CustomControl3)))
		End Sub
	End Class

	Public Class CustomControl4
		Inherits CustomControl1

		Shared Sub New()
			DefaultStyleKeyProperty.OverrideMetadata(GetType(CustomControl4), New FrameworkPropertyMetadata(GetType(CustomControl4)))
		End Sub
	End Class

	Public Class CustomControl5
		Inherits CustomControl1

		Shared Sub New()
			DefaultStyleKeyProperty.OverrideMetadata(GetType(CustomControl5), New FrameworkPropertyMetadata(GetType(CustomControl5)))
		End Sub
	End Class

	Public Class ThemeResourcesThemeKeyExtension
		Inherits ThemeKeyExtensionBase(Of ThemeResourcesThemeKeys)

		Public Overrides ReadOnly Property Assembly() As System.Reflection.Assembly
			Get
				Return If(TypeInTargetAssembly IsNot Nothing, TypeInTargetAssembly.Assembly, Me.GetType().Assembly)
			End Get
		End Property
	End Class
	Public Enum ThemeResourcesThemeKeys
		Background
		Foreground
		BorderBrush
		BorderThickness
		FontSize
	End Enum
End Namespace

File: CS/DXSample/CustomControls.cs
// Developer Express Code Central Example:
// How to implement the ThemeMananger theme support in custom controls
// 
// This is a sample project for the K18542 (http://www.devexpress.com/scid=K18542)
// KB article illustrating how to support themes in custom controls.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3524

using DevExpress.Xpf.Editors;
using DevExpress.Xpf.Utils.Themes;
using System.Windows;
using System.Windows.Controls;

namespace DXSample
{
	public class CustomControl1 : Control{
		static CustomControl1(){
			DefaultStyleKeyProperty.OverrideMetadata(typeof(CustomControl1), new FrameworkPropertyMetadata(typeof(CustomControl1)));
		}

        public string Text {
            get { return (string)GetValue(TextProperty); }
            set { SetValue(TextProperty, value); }
        }
        public static readonly DependencyProperty TextProperty =
            DependencyProperty.Register("Text", typeof(string), typeof(CustomControl1), new PropertyMetadata(null));
	}

    public class CustomControl2 : CustomControl1 {
        static CustomControl2() {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(CustomControl2), new FrameworkPropertyMetadata(typeof(CustomControl2)));
        }
    }

    public class CustomControl3 : CustomControl1 {
        static CustomControl3() {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(CustomControl3), new FrameworkPropertyMetadata(typeof(CustomControl3)));
        }
    }

    public class CustomControl4 : CustomControl1 {
        static CustomControl4() {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(CustomControl4), new FrameworkPropertyMetadata(typeof(CustomControl4)));
        }
    }

    public class CustomControl5 : CustomControl1 {
        static CustomControl5() {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(CustomControl5), new FrameworkPropertyMetadata(typeof(CustomControl5)));
        }
    }

    public class ThemeResourcesThemeKeyExtension : ThemeKeyExtensionBase<ThemeResourcesThemeKeys> {
        public ThemeResourcesThemeKeyExtension() {
        }
    }

    public enum ThemeResourcesThemeKeys {
        Background,
        Foreground,
        BorderBrush,
        BorderThickness,
        FontSize
    }

    public class CustomComboBoxEdit : ComboBoxEdit {

    }
}
